src/Polysemy/Final/Writer.hs

```diff
@@ -20,10 +29,53 @@ runWriterTVar tvar = runWriterTVarAction $ \o -> do
   writeTVar tvar $! s <> o
 {-# INLINE runWriterTVar #-}
 
+
+--------------------------------------------------------------------
+-- | Run a 'Writer' effect using atomic operations
+-- through final 'IO'.
+--
+-- Internally, this simply creates a new 'TVar', passes it to
+-- 'runWriterTVar', and then returns the result and the final value
+-- of the 'TVar'.
+--
+-- /Beware/: Effects that aren't interpreted in terms of 'IO'
+-- will have local state semantics in regards to 'Writer' effects
+-- interpreted this way. See 'Final'.
+writerToIOFinal :: (Member (Final IO) r, Monoid o)
+                => Sem (Writer o ': r) a
+                -> Sem r (o, a)
+writerToIOFinal sem = do
+  tvar <- embedFinal $ newTVarIO mempty
+  res  <- runWriterTVar tvar sem
+  end  <- embedFinal $ readTVarIO tvar
+  return (end, res)
+{-# INLINE writerToIOFinal #-}
+
+--------------------------------------------------------------------
+-- | Like 'writerToIOFinal'. but right-associates uses of '<>'.
+--
+-- This asymptotically improves performance if the time complexity of '<>'
+-- for the 'Monoid' depends only on the size of the first argument.
+--
+-- You should always use this instead of 'writerToIOFinal' if the monoid
+-- is a list, such as 'String'.
+--
+-- /Beware/: Effects that aren't interpreted in terms of 'IO'
+-- will have local state semantics in regards to 'Writer' effects
+-- interpreted this way. See 'Final'.
+writerToIOAssocRFinal :: (Member (Final IO) r, Monoid o)
+                      => Sem (Writer o ': r) a
+                      -> Sem r (o, a)
+writerToIOAssocRFinal =
```

Comment:
Here are some more interpreters I imagine will be useful. `writerToIOFinal` just removes the basic boilerplate usually attached to `-IORef` / `TVar` interpreters, and `writerToIOAssocRFinal` writes out the not-so-basic boilerplate needed to right-associate `tell`s for `writerToIOFinal`.

Don't worry, I won't add anything more to this pull request.

~~Unless you want a `Mask` effect~~